File: Lernwerk/Services/Calculator/CASEngine.swift
import Foundation
import UIKit
import WebKit

/// Runs Giac, compiled to WebAssembly, in a hidden web view and hands it the calculator's work. There is one engine
/// for the app: the calculator and the notes share its variables.
@MainActor
final class CASEngine: NSObject, ObservableObject {
    static let shared = CASEngine()

    enum State: Equatable {
        case idle, loading, ready
        case failed(String)
    }

    enum EngineError: LocalizedError {
        case unavailable(String)
        case badAnswer

        var errorDescription: String? {
            switch self {
            case let .unavailable(reason): return reason
            case .badAnswer: return "Der Rechenkern hat nicht verständlich geantwortet."
            }
        }
    }

    @Published private(set) var state: State = .idle
    private var webView: WKWebView?
    private var waiting: [CheckedContinuation<Void, Error>] = []
    private var timeout: DispatchWorkItem?
    /// Inputs that define variables and functions, made again whenever the engine starts.
    private var definitions: [(name: String, input: String)] = []
    private var degrees = false

    /// Starts loading Giac in the background, so it is ready by the time something needs it.
    func warmUp() {
        if state == .idle { start() }
    }

    private func start() {
        guard let page = Bundle.main.url(forResource: "cas", withExtension: "html") else {
            state = .failed("Der Rechenkern fehlt in dieser App.")
            return
        }
        state = .loading
        let configuration = WKWebViewConfiguration()
        configuration.userContentController.add(MessageProxy(engine: self), name: "cas")
        let webView = WKWebView(frame: CGRect(x: 0, y: 0, width: 1, height: 1), configuration: configuration)
        webView.navigationDelegate = self
        webView.isUserInteractionEnabled = false
        webView.alpha = 0.01
        // A web view outside every window may be paused by the system; a tiny one in the window keeps running.
        keyWindow?.addSubview(webView)
        webView.loadFileURL(page, allowingReadAccessTo: page.deletingLastPathComponent())
        self.webView = webView
        let work = DispatchWorkItem { [weak self] in
            MainActor.assumeIsolated {
                guard let self, self.state == .loading else { return }
                self.fail("Der Rechenkern ist nicht rechtzeitig gestartet.")
            }
        }
        timeout = work
        DispatchQueue.main.asyncAfter(deadline: .now() + 45, execute: work)
    }

    private var keyWindow: UIWindow? {
        let windows = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }.flatMap(\.windows)
        return windows.first { $0.isKeyWindow } ?? windows.first
    }

    fileprivate func receive(_ message: String) {
        guard message == "ready", state == .loading else { return }
        timeout?.cancel()
        Task { await restoreAndOpen() }
    }

    private func restoreAndOpen() async {
        if degrees { _ = try? await call("CAS.setDegrees(true); return 'ok'") }
        for definition in definitions {
            _ = try? await call("return CAS.evaluateJSON(input)", arguments: ["input": definition.input])
        }
        state = .ready
        let continuations = waiting
        waiting = []
        continuations.forEach { $0.resume() }
    }

    private func fail(_ reason: String) {
        state = .failed(reason)
        webView?.removeFromSuperview()
        webView = nil
        let continuations = waiting
        waiting = []
        continuations.forEach { $0.resume(throwing: EngineError.unavailable(reason)) }
    }

    private func ready() async throws {
        switch state {
        case .ready:
            return
        case .idle, .failed:
            // After a failure a new attempt may work, for example once memory is free again.
            start()
        case .loading:
            break
        }
        if case let .failed(reason) = state { throw EngineError.unavailable(reason) }
        try await withCheckedThrowingContinuation { continuation in
            waiting.append(continuation)
        }
    }

    /// Scripts always return a value: WebKit's bridge does not like an undefined result.
    private func call(_ script: String, arguments: [String: Any] = [:]) async throws -> Any {
        guard let webView else { throw EngineError.unavailable("Der Rechenkern läuft nicht.") }
        return try await withCheckedThrowingContinuation { continuation in
            webView.callAsyncJavaScript(script, arguments: arguments, in: nil, in: .page) { result in
                continuation.resume(with: result)
            }
        }
    }

    // Work

    func evaluate(_ input: String) async throws -> CASAnswer {
        try await ready()
        guard let json = try await call("return CAS.evaluateJSON(input)", arguments: ["input": input]) as? String else {
            throw EngineError.badAnswer
        }
        let answer = try JSONDecoder().decode(CASAnswer.self, from: Data(json.utf8))
        if answer.ok, let name = answer.assigns {
            definitions.removeAll { $0.name == name }
            definitions.append((name, input))
        }
        return answer
    }

    func plot(_ expressions: [String], xmin: Double, xmax: Double, samples: Int = 401) async throws -> CASPlot {
        try await ready()
        let list = String(data: try JSONEncoder().encode(expressions), encoding: .utf8) ?? "[]"
        let script = "return CAS.plotJSON(list, xmin, xmax, samples)"
        guard let json = try await call(script, arguments: ["list": list, "xmin": xmin, "xmax": xmax, "samples": samples]) as? String else {
            throw EngineError.badAnswer
        }
        return try JSONDecoder().decode(CASPlot.self, from: Data(json.utf8))
    }

    func setDegrees(_ value: Bool) async {
        degrees = value
        guard state == .ready else { return }
        _ = try? await call("CAS.setDegrees(value); return 'ok'", arguments: ["value": value])
    }

    /// The definitions and angle unit saved from earlier, made again before the first calculation.
    func restore(definitions saved: [CalculatorHistory.Definition], degrees: Bool) {
        definitions = saved.map { ($0.name, $0.input) }
        self.degrees = degrees
    }

    func forget(_ names: [String]) async {
        definitions.removeAll { names.contains($0.name) }
        guard state == .ready, let list = String(data: (try? JSONEncoder().encode(names)) ?? Data(), encoding: .utf8) else { return }
        _ = try? await call("CAS.forget(JSON.parse(list)); return 'ok'", arguments: ["list": list])
    }
}

extension CASEngine: WKNavigationDelegate {
    /// iOS ends the web content process when memory runs short; the next calculation starts Giac again.
    func webViewWebContentProcessDidTerminate(_ webView: WKWebView) {
        if state == .loading {
            fail("Der Rechenkern ist beim Laden abgestürzt.")
            return
        }
        self.webView?.removeFromSuperview()
        self.webView = nil
        state = .idle
    }

    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        fail("Der Rechenkern ließ sich nicht laden.")
    }

    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        fail("Der Rechenkern ließ sich nicht laden.")
    }
}

/// Keeps the web view from holding on to the engine.
@MainActor
private final class MessageProxy: NSObject, WKScriptMessageHandler {
    weak var engine: CASEngine?

    init(engine: CASEngine) {
        self.engine = engine
    }

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        engine?.receive(message.body as? String ?? "")
    }
}

File: android/app/src/main/java/de/maxifrz/lernwerk/calc/CasEngine.kt
package de.maxifrz.lernwerk.calc

import android.annotation.SuppressLint
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.webkit.JavascriptInterface
import android.webkit.RenderProcessGoneDetail
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json
import kotlin.coroutines.resume

/**
 * Runs Giac, compiled to WebAssembly, in a hidden WebView and hands it the calculator's work. There is one engine for
 * the app: the calculator and the notes share its variables. Everything that touches the WebView runs on the main
 * thread.
 */
class CasEngine private constructor(private val context: Context) {
    sealed interface State {
        data object Idle : State
        data object Loading : State
        data object Ready : State
        data class Failed(val reason: String) : State
    }

    class Unavailable(reason: String) : Exception(reason)

    private val _state = MutableStateFlow<State>(State.Idle)
    val state: StateFlow<State> = _state
    private val main = Handler(Looper.getMainLooper())
    private val json = Json { ignoreUnknownKeys = true }
    private var webView: WebView? = null
    private val waiting = mutableListOf<CompletableDeferred<Unit>>()
    /** Inputs that define variables and functions, made again whenever the engine starts. */
    private val definitions = mutableListOf<Pair<String, String>>()
    private var degrees = false
    private val timeout = Runnable { if (_state.value == State.Loading) fail("Der Rechenkern ist nicht rechtzeitig gestartet.") }

    /** Starts loading Giac in the background, so it is ready by the time something needs it. */
    fun warmUp() {
        main.post { if (_state.value == State.Idle) start() }
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun start() {
        _state.value = State.Loading
        val view = WebView(context)
        view.settings.javaScriptEnabled = true
        view.addJavascriptInterface(Bridge(), "CasBridge")
        view.webViewClient = object : WebViewClient() {
            override fun onRenderProcessGone(view: WebView, detail: RenderProcessGoneDetail): Boolean {
                // Android ends the renderer when memory runs short; the next calculation starts Giac again.
                if (_state.value == State.Loading) fail("Der Rechenkern ist beim Laden abgestürzt.") else discard(State.Idle)
                return true
            }

            override fun onReceivedError(view: WebView, request: WebResourceRequest, error: WebResourceError) {
                if (request.isForMainFrame) fail("Der Rechenkern ließ sich nicht laden.")
            }
        }
        view.loadUrl("file:///android_asset/cas.html")
        webView = view
        main.postDelayed(timeout, 45_000)
    }

    private inner class Bridge {
        @JavascriptInterface
        fun ready() {
            main.post { if (_state.value == State.Loading) onReady() }
        }
    }

    private fun onReady() {
        main.removeCallbacks(timeout)
        val view = webView ?: return
        val script = buildString {
            if (degrees) append("CAS.setDegrees(true);")
            definitions.forEach { (_, input) -> append("CAS.evaluateJSON(").append(quote(input)).append(");") }
            append("'ok'")
        }
        view.evaluateJavascript(script) {
            _state.value = State.Ready
            val deferreds = waiting.toList()
            waiting.clear()
            deferreds.forEach { it.complete(Unit) }
        }
    }

    private fun discard(next: State) {
        webView?.destroy()
        webView = null
        _state.value = next
    }

    private fun fail(reason: String) {
        main.removeCallbacks(timeout)
        discard(State.Failed(reason))
        val deferreds = waiting.toList()
        waiting.clear()
        deferreds.forEach { it.completeExceptionally(Unavailable(reason)) }
    }

    private suspend fun ready() = withContext(Dispatchers.Main) {
        when (_state.value) {
            State.Ready -> return@withContext
            State.Idle, is State.Failed -> start()
            State.Loading -> Unit
        }
        (_state.value as? State.Failed)?.let { throw Unavailable(it.reason) }
        val deferred = CompletableDeferred<Unit>()
        waiting += deferred
        deferred.await()
    }

    private fun quote(text: String) = json.encodeToString(String.serializer(), text)

    /** Runs a script that ends in a string and returns that string. */
    private suspend fun call(script: String): String = withContext(Dispatchers.Main) {
        val view = webView ?: throw Unavailable("Der Rechenkern läuft nicht.")
        val raw = suspendCancellableCoroutine { continuation ->
            view.evaluateJavascript(script) { continuation.resume(it) }
        }
        // evaluateJavascript hands back the value as JSON: a string arrives quoted.
        if (raw == null || raw == "null") throw Unavailable("Der Rechenkern hat nicht geantwortet.")
        json.decodeFromString(String.serializer(), raw)
    }

    suspend fun evaluate(input: String): CasAnswer {
        ready()
        val answer = json.decodeFromString(CasAnswer.serializer(), call("CAS.evaluateJSON(${quote(input)})"))
        val name = answer.assigns
        if (answer.ok && name != null) {
            withContext(Dispatchers.Main) {
                definitions.removeAll { it.first == name }
                definitions += name to input
            }
        }
        return answer
    }

    suspend fun plot(expressions: List<String>, xmin: Double, xmax: Double, samples: Int = 401): CasPlot {
        ready()
        val list = json.encodeToString(ListSerializer(String.serializer()), expressions)
        return json.decodeFromString(CasPlot.serializer(), call("CAS.plotJSON(${quote(list)}, $xmin, $xmax, $samples)"))
    }

    suspend fun setDegrees(value: Boolean) {
        withContext(Dispatchers.Main) { degrees = value }
        if (_state.value == State.Ready) runCatching { call("CAS.setDegrees($value); 'ok'") }
    }

    /** The definitions and angle unit saved from earlier, made again before the first calculation. */
    fun restore(saved: List<CalculatorHistory.Definition>, degrees: Boolean) {
        main.post {
            definitions.clear()
            definitions += saved.map { it.name to it.input }
            this.degrees = degrees
        }
    }

    suspend fun forget(names: List<String>) {
        withContext(Dispatchers.Main) { definitions.removeAll { it.first in names } }
        if (_state.value == State.Ready) {
            val list = json.encodeToString(ListSerializer(String.serializer()), names)
            runCatching { call("CAS.forget(JSON.parse(${quote(list)})); 'ok'") }
        }
    }

    companion object {
        @Volatile
        private var instance: CasEngine? = null

        fun get(context: Context): CasEngine =
            instance ?: synchronized(this) { instance ?: CasEngine(context.applicationContext).also { instance = it } }
    }
}
